File: java/problems/construct-binary-tree-from-inorder-and-postorder-traversal.java
class Solution {
  int x = 0;

  public TreeNode buildTree(int[] inorder, int[] postorder) {
    x = postorder.length - 1;
    return Solve(0, inorder.length - 1, inorder, postorder);
  }

  TreeNode Solve(int i, int j, int[] inorder, int[] postorder) {
    if (i > j)
      return null;

    int val = postorder[x];
    x -= 1;
    if (i == j) {
      return new TreeNode(val);
    }

    int c = i;
    while (inorder[c] != val)
      ++c;

    var node = new TreeNode(val);
    node.right = Solve(c + 1, j, inorder, postorder);
    node.left = Solve(i, c - 1, inorder, postorder);

    return node;
  }
}

File: python/problems/construct-binary-tree-from-inorder-and-postorder-traversal.py
from local_stuff import *


class Solution:
  def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:

    def solve(i, j):
      if i > j: return None
      val = postorder.pop()
      if i == j: return TreeNode(val)

      c = i
      while inorder[c] != val:
        c += 1

      node = TreeNode(val)
      node.right = solve(c + 1, j)
      node.left = solve(i, c - 1)

      return node

    return solve(0, len(inorder) - 1)


inorder = [9,3,15,20,7]
postorder = [9,15,7,20,3]

t = Solution().buildTree(inorder, postorder)
print(serializeTree(t))
